torchmetrics/wrappers/bootstrapping.py

```diff
@@ -0,0 +1,156 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from copy import deepcopy
+from typing import Any, Callable, List, Optional, Union
+
+import torch
+from torch import Tensor, nn
+
+from torchmetrics.metric import Metric
+from torchmetrics.utilities import apply_to_collection
+from torchmetrics.utilities.imports import _TORCH_GREATER_EQUAL_1_7
+
+
+def _bootstrap_sampler(
+    tensor: Tensor, size: Optional[int] = None, generator: Optional[torch.Generator] = None
+) -> Tensor:
+    """ Resample a tensor along its first dimension with replacement
+    Args:
+        tensor: tensor to resample
+        size: number of samples in new tensor. Defauls to same size as input tensor
+        generator: a instance of ``torch.Generator`` that controls the sampling
+
+    Returns:
+        resampled tensor
+
+    """
+    if size is None:
+        size = tensor.shape[0]
+    idx = torch.multinomial(
+        torch.ones(tensor.shape[0], device=tensor.device),
```

Comment:
I'm actually not sure if this is completely correct. For real bootstrap, we want to resample with replacement over the entire dataset, not on the batch level. After the correct bootstrap, it's possible that elements from this batch are not present in the result at all or present with more than 1 occurence.

I suspect the right formula will be attempting to model the number of times each individual sample needs to be present [0, inf)

Reply:
@maximsch2 thanks for raising this issue, I had not thought about it. I think it will be hard to implement "real" bootstrapping in batched version as we have no idea when to stop sampling from the current batch. However, I will try to think a bit more about this.
Else it should be stated in the docstring that this only perform "pseudo"-bootstrapping.

Reply:
@maximsch2 I added `sampling_strategy` argument and default to doing `poisson` sampling. As described here https://en.wikipedia.org/wiki/Bootstrapping_(statistics)#Poisson_bootstrap sampling using `Poisson(1)` to determine the number of times each sample should be included, will approximate the true bootstrapped distribution in the limit.

Reply:
@SkafteNicki, this looks like the right approach, thanks for adding this!